src/basic/REDUCE_STRUCT-Cuda.cpp

```diff
@@ -0,0 +1,192 @@
+  __syncthreads();
+
+  for ( i = blockDim.x / 2; i > 0; i /= 2 ) {
+    if ( threadIdx.x < i ) {
+	  //x
+      pxsum[ threadIdx.x ] += pxsum[ threadIdx.x + i ];
+      pxmin[ threadIdx.x ] = RAJA_MIN( pxmin[ threadIdx.x ], pxmin[ threadIdx.x + i ] );
+      pxmax[ threadIdx.x ] = RAJA_MAX( pxmax[ threadIdx.x ], pxmax[ threadIdx.x + i ] );
+	  //y
+      pysum[ threadIdx.x ] += pysum[ threadIdx.x + i ];
+      pymin[ threadIdx.x ] = RAJA_MIN( pymin[ threadIdx.x ], pymin[ threadIdx.x + i ] );
+      pymax[ threadIdx.x ] = RAJA_MAX( pymax[ threadIdx.x ], pymax[ threadIdx.x + i ] );
+
+    }
+     __syncthreads();
+  }
+
+// serialized access to shared data;
+  if ( threadIdx.x == 0 ) {
+    RAJA::atomicAdd<RAJA::cuda_atomic>( xsum, pxsum[ 0 ] );
+    RAJA::atomicMin<RAJA::cuda_atomic>( xmin, pxmin[ 0 ] );
+    RAJA::atomicMax<RAJA::cuda_atomic>( xmax, pxmax[ 0 ] );
+
+    RAJA::atomicAdd<RAJA::cuda_atomic>( xsum, pysum[ 0 ] );
+    RAJA::atomicMin<RAJA::cuda_atomic>( ymin, pymin[ 0 ] );
+    RAJA::atomicMax<RAJA::cuda_atomic>( ymax, pymax[ 0 ] );
+  }
+}
+
+template < size_t block_size >
+void REDUCE_STRUCT::runCudaVariantImpl(VariantID vid)
+{
+  const Index_type run_reps = getRunReps();
+  const Index_type ibegin = 0;
+  const Index_type iend = getActualProblemSize();
+
+  REDUCE_STRUCT_DATA_SETUP;
+
+  if ( vid == Base_CUDA ) {
+
+    REDUCE_STRUCT_DATA_SETUP_CUDA;
+
+    Real_ptr mem; //xcenter,xmin,xmax,ycenter,ymin,ymax
+    allocCudaDeviceData(mem,6);
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      cudaErrchk(cudaMemsetAsync(mem, 0.0, 6*sizeof(Real_type)));  
+      const size_t grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+                                                            
+      reduce_struct<block_size><<<grid_size, block_size,
+                                  6*sizeof(Real_type)*block_size>>>(
+        particles.x, particles.y,
+        mem, mem+1, mem+2,    // xcenter,xmin,xmax
+        mem+3, mem+4, mem+5,  // ycenter,ymin,ymax
+        particles.N);
+      cudaErrchk( cudaGetLastError() );
+
+      Real_type lmem[6]={0.0, 0.0, 0.0, 0.0, 0.0, 0.0};
+      Real_ptr plmem = &lmem[0];
+      getCudaDeviceData(plmem, mem, 6);
+
+      particles.SetCenter(lmem[0]/particles.N,lmem[3]/particles.N);
+      particles.SetXMin(lmem[1]);
+      particles.SetXMax(lmem[2]);
+      particles.SetYMin(lmem[4]);
+      particles.SetYMax(lmem[5]);
+      m_particles=particles;
+
+    }
+    stopTimer();
+
+    REDUCE_STRUCT_DATA_TEARDOWN_CUDA;
+
+    deallocCudaDeviceData(mem);
+
+  } else if ( vid == RAJA_CUDA ) {
+
+    REDUCE_STRUCT_DATA_SETUP_CUDA;
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      RAJA::ReduceSum<RAJA::cuda_reduce, Real_type> xsum(0.0);
+      RAJA::ReduceSum<RAJA::cuda_reduce, Real_type> ysum(0.0);
+      RAJA::ReduceMin<RAJA::cuda_reduce, Real_type> xmin(0.0); 
+      RAJA::ReduceMin<RAJA::cuda_reduce, Real_type> ymin(0.0);
```

Comment:
Should the Mins be initialized to `std::numeric_limits<Int_type>::max();` ? like above? Similarly for the Max, but with ::min() ?

Reply:
@artv3 good catch. all reduction objects and variables should be initialized using the 'm_*' class members that I described in a comment up the page.